### PR TITLE
Update Unix dependencies installation script

### PR DESCRIPTION
**Pull request by am11**
* Use version-less llvm toolchain on Debian/Ubuntu (https://github.com/dotnet/runtime/pull/74811).
* Allow case insensitive OS name (`eng/install-native-dependencies.sh macos # or osx`).
* Check if distro is Debian or its derivative (`apt` is a Debian thing).
* Remove `sudo` to make it work with root (e.g. in a docker container); non-root will get a (readable) error from the system, when higher privileges are needed.
* Fix shellcheck errors and lower the dependency to Unix/Bourne shell script (we were using mixed syntax, now it is unified).

### CODE REVIEW

File: eng/install-native-dependencies.sh [hunk 1]
```diff
@@ -13,37 +15,35 @@
...
+
+    localedef -i en_US -c -f UTF-8 -A /usr/share/locale/locale.alias en_US.UTF-8
+elif [ "$os" = "maccatalyst" ] || [ "$os" = "osx" ] || [ "$os" = "macos" ] || [ "$os" = "tvos" ] || [ "$os" = "ios" ]; then
+    echo "Installed xcode version: $(xcode-select -p)"
 
     if [ "$3" = "azDO" ]; then
         # workaround for old osx images on hosted agents
         # piped in case we get an agent without these values installed
-        if ! brew_output="$(brew uninstall openssl@1.0.2t 2>&1 >/dev/null)"; then
+        if ! brew uninstall openssl@1.0.2t >/dev/null 2>&1; then
```
Comment (ViktorHofer) [inline eng/install-native-dependencies.sh:37]: Do you know if this issue still applies to the current osx images on our hosted agents?

File: eng/install-native-dependencies.sh [hunk 2]
```diff
@@ -13,37 +15,35 @@
...
+fi
 
-    echo "Installed xcode version: `xcode-select -p`"
+if [ "$os" = "linux" ] && { [ "$ID" = "debian" ] || [ "$ID_LIKE" = "debian" ]; }; then
+    apt update
+
+    apt install -y build-essential gettext locales cmake llvm clang lldb liblldb-dev libunwind8-dev libicu-dev liblttng-ust-dev \
+        libssl-dev libkrb5-dev libnuma-dev
+
+    localedef -i en_US -c -f UTF-8 -A /usr/share/locale/locale.alias en_US.UTF-8
```
Comment (ViktorHofer) [inline eng/install-native-dependencies.sh:30]: This wasn't there before, right? Why is this required?

Comment (am11) [reply eng/install-native-dependencies.sh:30]: Yes, before locales and libz-dev were missing and libcurl-openssl-dev was unnecessary. So it was simply outdated.

e.g. given a new environment:

```sh
docker run --rm -v $(pwd):/runtime ubuntu sh -c \
  '/runtime/eng/install-native-dependencies.sh linux && /runtime/build.sh libs.native'
```

without any locales gives a lot of errors like this:

```
Build succeeded.

bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/runtime/src/native/libs/build-native.proj]
bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/runtime/src/native/libs/build-native.proj]
bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/runtime/src/native/libs/build-native.proj]
bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/runtime/src/native/libs/build-native.proj]
bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/runtime/src/native/libs/build-native.proj]
bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/runtime/src/native/libs/build-native.proj]
bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/runtime/src/native/libs/build-native.proj]
bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/runtime/src/native/libs/build-native.proj]
bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/runtime/src/native/libs/build-native.proj]
bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/runtime/src/native/libs/build-native.proj]
bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/runtime/src/native/libs/build-native.proj]
    0 Warning(s)
    11 Error(s)

Time Elapsed 00:02:57.08
Build failed with exit code 1. Check errors above.
```

File: eng/install-native-dependencies.sh (hunk 1 above)
Comment (am11) [reply eng/install-native-dependencies.sh:37]: Not sure if the workaround is really needed, but `azDO` argument is in use:

```sh
$ git grep install-native-dependencies eng/pipelines 

eng/pipelines/common/global-build-job.yml:      - script: $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }} ${{ parameters.archType }} azDO
eng/pipelines/common/templates/runtimes/build-test-job.yml:      - script: $(Build.SourcesDirectory)/eng/install-native-dependencies.sh $(osGroup) ${{ parameters.archType }} azDO
eng/pipelines/coreclr/templates/build-jit-job.yml:      - script: $(Build.SourcesDirectory)/eng/install-native-dependencies.sh $(osGroup) ${{ parameters.archType }} azDO
eng/pipelines/coreclr/templates/build-job.yml:      - script: $(Build.SourcesDirectory)/eng/install-native-dependencies.sh $(osGroup) ${{ parameters.archType }} azDO
eng/pipelines/installer/jobs/build-job.yml:        - script: $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }} ${{ parameters.archType }} azDO
eng/pipelines/libraries/build-job.yml:          - script: $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }} ${{ parameters.archType }} azDO
eng/pipelines/mono/templates/build-job.yml:      - script: $(Build.SourcesDirectory)/eng/install-native-dependencies.sh $(osGroup) ${{ parameters.archType }} azDO
eng/pipelines/mono/templates/generate-offsets.yml:      - script: $(Build.SourcesDirectory)/eng/install-native-dependencies.sh $(osGroup) ${{ parameters.archType }} azDO
```

File: eng/install-native-dependencies.sh [hunk 3]
```diff
@@ -13,37 +15,35 @@
...
+if [ -e /etc/os-release ]; then
+    . /etc/os-release
+fi
 
-    echo "Installed xcode version: `xcode-select -p`"
+if [ "$os" = "linux" ] && { [ "$ID" = "debian" ] || [ "$ID_LIKE" = "debian" ]; }; then
+    apt update
+
+    apt install -y build-essential gettext locales cmake llvm clang lldb liblldb-dev libunwind8-dev libicu-dev liblttng-ust-dev \
+        libssl-dev libkrb5-dev libnuma-dev
```
Comment (am11) [inline eng/install-native-dependencies.sh:28]:
```suggestion
        libssl-dev libkrb5-dev libnuma-dev libz-dev
```

File: eng/install-native-dependencies.sh (hunk 1 above)
Comment (akoeplinger) [reply eng/install-native-dependencies.sh:37]: As far as I know this issue no longer happens on the macOS hosted agents so we can try removing it. I don't have a preference for whether you want to do it as part of this PR or separately

File: eng/install-native-dependencies.sh [hunk 4]
```diff
@@ -13,37 +15,35 @@
...
+if [ -e /etc/os-release ]; then
+    . /etc/os-release
+fi
 
-    echo "Installed xcode version: `xcode-select -p`"
+if [ "$os" = "linux" ] && { [ "$ID" = "debian" ] || [ "$ID_LIKE" = "debian" ]; }; then
+    apt update
+
+    apt install -y build-essential gettext locales cmake llvm clang lldb liblldb-dev libunwind8-dev libicu-dev liblttng-ust-dev \
+        libssl-dev libkrb5-dev libnuma-dev libz-dev
```
Comment (akoeplinger) [inline eng/install-native-dependencies.sh:28]: the [docs](https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md#toolchain-setup) suggest zlib1g-dev instead:
 
```suggestion
        libssl-dev libkrb5-dev libnuma-dev zlib1g-dev
```

File: eng/install-native-dependencies.sh (hunk 2 above)
Comment (akoeplinger) [reply eng/install-native-dependencies.sh:30]: Looks good to me, it is what we use in some Docker setup scripts like https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/db4fcc31b3fe5315bc0ae7f02e9e34cd16bbe102/src/debian/11/gcc12/Dockerfile#L24

And we're only using this script on macOS so it doesn't impact our builds anyway.

**Review by akoeplinger — APPROVED** [review]

File: eng/install-native-dependencies.sh (hunk 4 above)
Comment (am11) [reply eng/install-native-dependencies.sh]: [libz-dev](https://packages.ubuntu.com/jammy/libz-dev) installs zlib1g-dev. i like libz-dev because it shares naming convention with every other debian dev packages.

**Review by ViktorHofer — APPROVED** [review]
Thanks a ton 🤘

**Review by akoeplinger — APPROVED** [review]
